src/ucp/core/ucp_context.h

```diff
@@ -441,45 +441,64 @@ static UCS_F_ALWAYS_INLINE int ucp_memory_type_cache_is_empty(ucp_context_h cont
             !context->memtype_cache->pgtable.num_regions);
 }
 
-static UCS_F_ALWAYS_INLINE ucs_memory_type_t
-ucp_memory_type_detect(ucp_context_h context, const void *address, size_t length)
+static UCS_F_ALWAYS_INLINE void
+ucp_memory_info_set_host(ucs_memory_info_t *mem_info)
+{
+    mem_info->type    = UCS_MEMORY_TYPE_HOST;
+    mem_info->sys_dev = UCS_SYS_DEVICE_ID_UNKNOWN;
+}
+
+static UCS_F_ALWAYS_INLINE void
+ucp_memory_detect(ucp_context_h context, const void *address, size_t length,
+                  ucs_memory_info_t *mem_info)
 {
-    ucs_memory_type_t mem_type;
     ucs_status_t status;
 
     if (ucs_likely(context->num_mem_type_detect_mds == 0)) {
-        return UCS_MEMORY_TYPE_HOST;
+        goto out_host_mem;
     }
 
     if (ucs_likely(context->memtype_cache != NULL)) {
         if (!context->memtype_cache->pgtable.num_regions) {
-            return UCS_MEMORY_TYPE_HOST;
+            goto out_host_mem;
         }
 
         status = ucs_memtype_cache_lookup(context->memtype_cache, address,
-                                          length, &mem_type);
-        if (status != UCS_OK) {
+                                          length, mem_info);
+        if (ucs_likely(status != UCS_OK)) {
             ucs_assert(status == UCS_ERR_NO_ELEM);
-            return UCS_MEMORY_TYPE_HOST;
+            goto out_host_mem;
         }
 
-        if (mem_type != UCS_MEMORY_TYPE_LAST) {
-            return mem_type;
+        if ((mem_info->type != UCS_MEMORY_TYPE_LAST) &&
```

Comment:
should be `mem_info->type != UCS_MEMORY_TYPE_UNKNWON` right?

Reply:
fixed